File: main/Exp.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Interop
{
    public class Exp
    {
    }

    public class Id : Exp
    {
        public string name;

        public Id(string name)
        {
            this.name = name;
        }
    }

    public class Const : Exp
    {
        public int c;

        public Const(int n)
        {
            this.c = n;
        }
    }

    public class IAdd : Exp
    {
        public Exp left;
        public Exp right;

        public IAdd(Exp left, Exp right)
        {
            this.left = left;
            this.right = right;
        }
    }
}

File: driver/exp.cpp
#include <memory>
#include <stack>
#include <vector>
#include <ostream>
#include "types.h"
#include "RekoInterfaces.h"
#include "exp.h"
#include "stmt.h"
#include "NativeFactory.h"

void BinOp::write(std::wostream & stm)
{
	stm << '(';
	this->Left->write(stm);
	wchar_t * s;
	switch (this->Operator)
	{
	case PrimitiveOp::IAdd: s = L" + "; break;
	case PrimitiveOp::ISub: s = L" - "; break;
	default: s = L"?unknown?";
	}
	stm << s;
	this->Right->write(stm);
	stm << ')';
}
